python/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,68 @@
+# cuML python developer guide
+This document summarizes rules and best practices for contributions to the cuML python component of rapidsai/cuml. This is a living document and contributions for clarifications or fixes and issue reports are highly welcome.
+
+## General
+Please start by reading:
+1. [CONTRIBUTING.md](../CONTRIBUTING.md).
+2. [C++ DEVELOPER_GUIDE.md](../cuML/DEVELOPER_GUIDE.md)
+3. [Python cuML README.md](README.md)
+
+## Thread safety
+Refer to the section on thread safety in [C++ DEVELOPER_GUIDE.md](../cuML/DEVELOPER_GUIDE.md#thread-safety)
+
+## Coding style
+1. [PEP8](https://www.python.org/dev/peps/pep-0008)
+2. [sklearn coding guidelines](https://scikit-learn.org/stable/developers/contributing.html#coding-guidelines)
+
+## Creating class for a new ML algo
+1. Make sure that this algo has been implemented in the C++ side. Refer to [C++ DEVELOPER_GUIDE.md](../cuML/DEVELOPER_GUIDE.md) for guidelines on developing in C++.
+2. Create a corresponding algoName.pyx file inside `python/cuml` folder.
+3. Note that the folder structure inside here should reflect sklearn's. Example, `pca.pyx` should be kept inside the `decomposition` sub-folder of `python/cuml`.
+4. We try to match the corresponding scikit-learn's interface as closely as possible. Refer to their [developer guide](https://scikit-learn.org/stable/developers/contributing.html#apis-of-scikit-learn-objects) on API design of sklearn objects for details.
+5. Always make sure to have your class inherit from `cuml.common.base.Base` class as your parent/ancestor.
+
+## Error handling
+If you are trying to call into cuda runtime APIs inside `cuml.common.cuda`, in case of any errors, they'll raise a `cuml.common.cuda.CudaRuntimeError`. For example:
+```python
+from cuml.common.cuda import Stream, CudaRuntimeError
+try:
+    s = Stream()
+    s.sync
+except CudaRuntimeError as cre:
+    print("Cuda Error! '%s'" % str(cre))
+```
```

Comment:
Adding a note of Cython catching C++ exception catching easily with except (i.e. `cumlHandle() except +`)